### PR TITLE
TST: add test for Index.where when needing to cast to object dtype

### LINKED ISSUE #32413: Object dtype gives ValueError in _maybe_cast_with_dtype
#### 

```python
import pandas as pd
import numpy as np

frame = pd.DataFrame(index=pd.Index([1, np.nan]))
fixed_index = frame.index.where(frame.index.notna(), 'a' + frame.reset_index().index.astype(str))

Traceback (most recent call last):
  File "<input>", line 4, in <module>
  File "C:\Users\Miniconda3\envs\testbed_dtype\lib\site-packages\pandas\core\indexes\base.py", line 3791, in where
    return self._shallow_copy_with_infer(values, dtype=dtype)
  File "C:\Users\Miniconda3\envs\testbed_dtype\lib\site-packages\pandas\core\indexes\base.py", line 567, in _shallow_copy_with_infer
    return Index(values, **attributes)
  File "C:\Users\Miniconda3\envs\testbed_dtype\lib\site-packages\pandas\core\indexes\base.py", line 385, in __new__
    data = _maybe_cast_with_dtype(data, dtype, copy)
  File "C:\Users\Miniconda3\envs\testbed_dtype\lib\site-packages\pandas\core\indexes\base.py", line 5457, in _maybe_cast_with_dtype
    data = data.astype(dtype)
ValueError: could not convert string to float: 'a1'
```
#### Problem description

Data is being cast to incorrect dtype based on inferred dtype and gives a ValueError.
Issue was not present in pandas==0.25.3

#### Expected Output
With pandas==0.25.3

fixed_index
Index([1.0, 'a1'], dtype='object')

#### Output of ``pd.show_versions()``
python           : 3.7.6.final.0
pandas           : 1.0.1
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 45.2.0.post20200210
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : 1.3.0
xlsxwriter       : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by cgarciae**
- [x] closes #32413
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry: Adds a basic test for #32413.

Hey! This is my first PR with actual code: added a very basic test to validate #32413. 

Questions:
* What is the best place for this test? (currently put it in `pandas/tests/indexes/test_indexing.py`)
* Should I try to create another test that uses the `index` input argument?

### CODE REVIEW

File: pandas/tests/indexes/test_indexing.py [hunk 1]
```diff
@@ -248,6 +249,20 @@ def test_putmask_with_wrong_mask(self, index):
             index.putmask("foo", fill)
 
 
+class TestCasting:
+    def test_maybe_cast_with_dtype(self):
```
Comment (mroeschke) [inline pandas/tests/indexes/test_indexing.py:253]: This test is better suited in `pandas/tests/indexes/base_class/test_where.py`

File: pandas/tests/indexes/test_indexing.py [hunk 2]
```diff
@@ -248,6 +249,20 @@ def test_putmask_with_wrong_mask(self, index):
...
+        # https://github.com/pandas-dev/pandas/issues/32413
+        frame = DataFrame(index=Index([1, np.nan]))
+
+        cond = frame.index.notna()
+        other = "a" + frame.reset_index().index.astype(str)
+
+        fixed_index = frame.index.where(cond, other)
+
+        assert (other == Index(["a0", "a1"])).all()
```
Comment (mroeschke) [inline pandas/tests/indexes/test_indexing.py:262]: Generally should be asserted using `tm.assert_index_equal(result, expected)`

File: pandas/tests/indexes/test_indexing.py [hunk 3]
```diff
@@ -248,6 +249,20 @@ def test_putmask_with_wrong_mask(self, index):
...
 
+class TestCasting:
+    def test_maybe_cast_with_dtype(self):
+        # https://github.com/pandas-dev/pandas/issues/32413
+        frame = DataFrame(index=Index([1, np.nan]))
+
+        cond = frame.index.notna()
+        other = "a" + frame.reset_index().index.astype(str)
+
+        fixed_index = frame.index.where(cond, other)
```
Comment (jorisvandenbossche) [inline pandas/tests/indexes/test_indexing.py:260]: I know this was the original reproducer, but for the actual `where` call, we don't need to DataFrame. You can directly create the index and work with that. Something like (untested):


```suggestion
        index = Index([1, np.nan]))

        cond = index.notna()
        other = "a" + Index(rang(2)).astype(str)

        fixed_index = index.where(cond, other)
```

**Review by jorisvandenbossche — COMMENTED** [review]
@cgarciae Thanks for the PR!

(side-note: for future reference, try to give your PRs a descriptive title; only reference the issue number is not that useful (since we don't know by heart what those numbers are about ;))

**Comment by cgarciae** [comment]
Thanks @jorisvandenbossche and @mroeschke for the feedback! I'll clean the PR a bit :)

File: pandas/tests/indexes/test_indexing.py (hunk 3 above)
Comment (cgarciae) [reply pandas/tests/indexes/test_indexing.py]: Worked great, thanks!

File: pandas/tests/indexes/test_indexing.py (hunk 2 above)
Comment (cgarciae) [reply pandas/tests/indexes/test_indexing.py]: Thanks! I've changed the test to use this.

File: pandas/tests/indexes/test_indexing.py (hunk 1 above)
Comment (cgarciae) [reply pandas/tests/indexes/test_indexing.py:252]: The main issue was the casting so I was not sure, I could move it anyway.

Comment (mroeschke) [reply pandas/tests/indexes/test_indexing.py:252]: Thanks, Generally we group tests by the method calls on the particular objects (`Index.where`)

**Review by jorisvandenbossche — COMMENTED** [review]
As a last comment, could you still move the test? 
I think `tests/indexes/numeric/test_indexing.py` might be best suites, there is a `TestWhere` class

**Comment by cgarciae** [comment]
@jorisvandenbossche moved the test there!

**Review by mroeschke — APPROVED** [review]

File: pandas/tests/indexes/numeric/test_indexing.py [hunk 4]
```diff
@@ -376,6 +376,18 @@ def test_where(self, klass, index):
         result = index.where(klass(cond))
         tm.assert_index_equal(result, expected)
 
+    def test_maybe_cast_with_dtype(self):
```
Comment (jbrockmendel) [inline pandas/tests/indexes/numeric/test_indexing.py:379]: can you give this a name of the form `test_where_[...]` ideally more desc

**Comment by jbrockmendel** [comment]
the underlying problem in #32413 was at a lower level IIRC.  should we have a test on astype or a constructor?

File: pandas/tests/indexes/numeric/test_indexing.py [hunk 5]
```diff
@@ -376,6 +376,18 @@ def test_where(self, klass, index):
...
+        # https://github.com/pandas-dev/pandas/issues/32413
+        index = Index([1, np.nan])
+
+        cond = index.notna()
+        other = "a" + Index(range(2)).astype(str)
+
+        fixed_index = index.where(cond, other)
+
+        tm.assert_index_equal(other, Index(["a0", "a1"]))
```
Comment (jreback) [inline pandas/tests/indexes/numeric/test_indexing.py:388]: can you use

result = 
expected = 
tm.assert_index_equal(result, expected)

Comment (jreback) [inline pandas/tests/indexes/numeric/test_indexing.py:388]: pls parameterize if you have multiple cases

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by simonjayhawkins** [comment]
@cgarciae can you address @jreback comments.

**Comment by jreback** [comment]
closing as stale

**Pull request closed without merging**